sec2-architecture.tex

```diff
@@ -36,28 +36,62 @@ \subsubsection{Functional Architecture}\label{functional-architecture}
-the services hold minimal non-persistent state so
-that loss of a single service instance has correspondingly minimal impact on
-users.
+using Kubernetes.  The underlying provisioning of compute resources is
+handled by the Provisioning and Deployment Service component of the LSST DMS.
+ The technology to be used for provisioning in this component is being
+evaluated and may vary between instances.  vSphere, HTCondor, SLURM, and
+OpenStack are all potential alternatives.  Provisioning is elastic: JupyterLab
+dynamically instantiates notebooks on demand, and the Portal and Web API
+services can be expanded or contracted as needed. This elasticity, together
+with fault tolerance considerations, require that the services hold minimal
+non-persistent state so that loss of a single service instance has
+correspondingly minimal impact on users.
+
+The LSST Data Facility (LDF) will operate production and development clusters
+of the LSP services. The production LSP services are deployed onto the
+Kubernetes platform by LDF service production staff. The containers are
+deployed using images in local container repositories. There are no external
+dependencies so that in the case of network outages at a site containers can
+still be deployed, even if the Kubernetes infrastructure needs to be
+rebuilt. Batch production services are co-located with LSP service
+instances. Data Backbone endpoint services are set up locally to the Base and
+the LDF.
+
+Containers to be deployed to Kubernetes are built using the automated build
+system.  The containers are placed into local repositories on Kubernetes test
+systems vetted by security staff, and then deployed to the production
+container registry. Security and monitoring software are put in place to
+supervise the activity of the host systems and containers.
+
+Kubernetes releases happen in regular and frequent, intervals. LDF staff
+monitor these releases and evaluate the appropriateness of upgrading,
+balancing any new features we might be able to take advantage of against any
+disruption or change in users’ workflow.
+
+There are two types of updates to the software that will take place:  
+Kubernetes and LSP services.  Updates will be implemented on two test
+clusters. The first cluster will be used by system administration to test 
+configurations and system software updates.  The second is used for LSP 
+service development.
+
+Updates to Kubernetes will first be done on the system administration cluster
+to verify update procedures. After this has been successfully completed, the
+update will be done to the LSP development cluster, and regression tests for
+LSP services will take place.  After this has been verified,  Kubernetes will
+be upgraded on the production cluster.
+
+Updates to LSP services are first done on LSP service development cluster. 
+After all tests have passed, the LSP service updates will be done on the 
+production cluster.
```

Comment:
Again, does this mean that all production instances are updated at the same time?  I hope not.

Reply:
No, those are done individually, unless there's some kind of dependency between the services.  The Kubernetes is a bit different, since (usually) there can be more of a roll-out of components from old version to new version, on the fly.